### PR TITLE
Feat: enhanced horizontal scaling of app controller

### PR DESCRIPTION
**Pull request by fourierr**
Signed-off-by: Xiangbo Ma <maxiangboo@cmbchina.com>

### Description of your changes

Fixes https://github.com/kubevela/kubevela/issues/4596

Add sharding for app controller, and legacy app cr will be processed by sharding-1. 
I have:

- [x] Read and followed KubeVela's [contribution process](https://github.com/kubevela/kubevela/blob/master/contribute/create-pull-request.md).
- [ ] [Related Docs](https://github.com/kubevela/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### CODE REVIEW

File: cmd/core/main.go [hunk 1]
```diff
@@ -98,6 +98,8 @@ func main() {
 		"Enable leader election for controller manager. Enabling this will ensure there is only one active controller manager.")
 	flag.StringVar(&leaderElectionNamespace, "leader-election-namespace", "",
 		"Determines the namespace in which the leader election configmap will be created.")
+	flag.StringVar(&leaderElectionName, "leader-election-name", "sharding-1",
```
Comment (wonderflow) [inline cmd/core/main.go:101]: 1. by default it should be empty.
2. `leader-election-name` -> `app-sharding-key` \c @Somefive  any good suggestion?

File: cmd/core/main.go [hunk 2]
```diff
@@ -230,7 +232,7 @@ func main() {
 	}
 	ctrl.SetLogger(klogr.New())
 
-	leaderElectionID := util.GenerateLeaderElectionID(types.KubeVelaName, controllerArgs.IgnoreAppWithoutControllerRequirement)
+	leaderElectionID := util.GenerateLeaderElectionID(types.KubeVelaName+"-"+leaderElectionName, controllerArgs.IgnoreAppWithoutControllerRequirement)
```
Comment (wonderflow) [inline cmd/core/main.go:235]: I wonder if it will cause incompatible?

**Review by wonderflow — CHANGES_REQUESTED** [review]
1. how this PR be tested?
2. please also add some test cases
3. please add docs for this feature

File: cmd/core/main.go (hunk 2 above)
Comment (fourierr) [reply cmd/core/main.go:235]: I think it's a leader-election configmap name and be compatible.

File: cmd/core/main.go (hunk 1 above)
Comment (fourierr) [reply cmd/core/main.go:101]: > 1. by default it should be empty.

I think for a while, if default is empty, and we want to have three sharding, the sharding will contains "", "sharding-1", "sharding-2", it's a little strange. I'm not sure whether I'm proper.

**Comment by fourierr** [comment]
> 1. how this PR be tested?
> 2. please also add some test cases
> 3. please add docs for this feature

I plan to deploy app controller with `sharding-1`, and create three app cr, 
- the first cr was default and no mark, 
- the second cr was marked with `sharding-1`. 
- the third cr was marked with `sharding-2`. 
The expected result is:
- the first cr was processed successful,
- the second cr was also processed successful,
- the third cr has not been processed.
How about this testing case? @wonderflow

File: cmd/core/main.go (hunk 1 above)
Comment (fourierr) [reply cmd/core/main.go:101]: After thinking again, most users will only use one sharding. So It is better to default to empty.

**Comment by fourierr** [comment]
At first, I wanted to set multiple shardings through `helm -- set`, but in most cases, only one app controller was deployed at the beginning. Later, it was found that it could not be satisfied, and then the app controller was gradually added sharding. In this case, helm cannot implement it, so I plan to increase the number of shardings gradually through `vela sharding add`. How about `vela sharding add`?

**Comment by wonderflow** [comment]
The sharding controller only consumes those apps with sharding annotation. What's blocked on helm installation?

**Comment by Somefive** [comment]
I have some design questions for this PR. I understand that sharding could help make controller horizontally scalable. But technically, I think `controller-requirement` can already satisfy the need of sharding. The only problem is the shared leader election lock cause multiple controllers cannot work concurrently.

Therefore, I think it might not be necessary to add `app.oam.dev/sharding` for applications. It should be able to reuse `app.oam.dev/controller-version-require`.

**Comment by wonderflow** [comment]
> I have some design questions for this PR. I understand that sharding could help make controller horizontally scalable. But technically, I think `controller-requirement` can already satisfy the need of sharding. The only problem is the shared leader election lock cause multiple controllers cannot work concurrently.
> 
> Therefore, I think it might not be necessary to add `app.oam.dev/sharding` for applications. It should be able to reuse `app.oam.dev/controller-version-require`.

I don't think we should re-sue `app.oam.dev/controller-version-require` in this case, reasons:

1. Users can't understand the annotation, this is mainly for inner apiserver to use.
2. In some cases, it could be possible that we both these two Annotations, for example, a sharding key for some BU while there're two versions of controller to rollout.

**Comment by Somefive** [comment]
> > I have some design questions for this PR. I understand that sharding could help make controller horizontally scalable. But technically, I think `controller-requirement` can already satisfy the need of sharding. The only problem is the shared leader election lock cause multiple controllers cannot work concurrently.
> > Therefore, I think it might not be necessary to add `app.oam.dev/sharding` for applications. It should be able to reuse `app.oam.dev/controller-version-require`.
> 
> I don't think we should re-sue `app.oam.dev/controller-version-require` in this case, reasons:
> 
> 1. Users can't understand the annotation, this is mainly for inner apiserver to use.
> 2. In some cases, it could be possible that we both these two Annotations, for example, a sharding key for some BU while there're two versions of controller to rollout.

I see. Then the implementation is relatively reasonable for me.

**Comment by fourierr** [comment]
> The sharding controller only consumes those apps with sharding annotation. What's blocked on helm installation?
> 
> The default behavior is the default controller don't consume apps with sharding annotation.

I didn't think there was any blocks to helm.

File: cmd/core/main.go (hunk 2 above)
Comment (wonderflow) [reply cmd/core/main.go:235]: if leaderElectionName is empty, the "-" should be omitted

File: pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go [hunk 3]
```diff
@@ -622,3 +627,7 @@ func (r *Reconciler) matchControllerRequirement(app *v1beta1.Application) bool {
 	}
 	return true
 }
+
+func (r *Reconciler) matchAppSharding(object client.Object) bool {
+	return object.GetLabels()[oam.LabelAppSharding] == pflag.Lookup("leader-election-name").Value.String()
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go:632]: don't look up every time, the performance could be better

Comment (fourierr) [reply pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go:632]: > don't look up every time, the performance could be better

ok, i will fix it in recently

**Comment by Somefive** [comment]
Duplicated with #4705

**Pull request closed without merging**